gapic/samplegen/samplegen.py

```diff
@@ -102,6 +104,81 @@ class TransformedRequest:
     base: str
     single: Optional[AttributeRequestSetup]
     body: Optional[List[AttributeRequestSetup]]
+    pattern: Optional[str] = None
+
+    # Resource patterns look something like
+    # kingdom/{kingdom}/phylum/{phylum}/class/{class}
+    RESOURCE_RE = re.compile(r"\{([^}/]+)\}")
+
+    @classmethod
+    def build(cls, request_type, api_schema, base, attrs, is_resource_request):
+        """Build a TransformedRequest based on parsed inpu
+
+        Acts as a factory to hide complicated logic for resource based requests.
+
+        Args:
+            request_type (wrappers.MessageType): The method's request message type.
+            api_schema (api.API): The API schema (used for looking up other messages)
+            base (str): the name of the base field being set.
+            attrs [AttributeRequestSetup]: All the attributes (or fields) being set
+                                           for the base field.
+            is_resource_request (bool): Indicates whether the request describes a
+                                        constructed resource name path.
+        """
+
+        if not attrs[0].field:
+            return cls(base=base, body=None, single=attrs[0])
+        elif not is_resource_request:
+            return cls(base=base, body=attrs, single=None)
+        else:
+            # This is the tricky one.
+            # We need to determine whether the field is describing a valid resource,
+            # and if so, what its corresponding message type is.
+            # Then we need to find the pattern with parameters
+            # that exatcly matches the attrs, if one exists.
+            #
+            # It's a precondition that the base field is
+            # a valid field of the request message type.
+            resource_typestr = (request_type.
+                                fields[base].
+                                options.
+                                Extensions[resource_pb2.resource_reference].
+                                type)
+
+            resource_message_descriptor = next(
+                (msg.options.Extensions[resource_pb2.resource]
+                 for msg in api_schema.messages.values()
+                 if msg.options.Extensions[resource_pb2.resource].type == resource_typestr),
```

Comment:
Note: please also check [file-level annotations](https://github.com/googleapis/googleapis/blob/d9576d95b44f64fb0e3da4760adfc4a24fa1faab/google/api/resource.proto#L35-L39). Some resource type might not have a message.

You should also have a way to instantiate the five common resource names: 
- cloudresourcemanager.googleapis.com/Project
- cloudresourcemanager.googleapis.com/Organization
- cloudresourcemanager.googleapis.com/Folder
- cloudbilling.googleapis.com/BillingAccount
- locations.googleapis.com/Location

Their patterns: https://github.com/googleapis/gapic-generator/blob/master/src/test/java/com/google/api/codegen/testsrc/common/common_resources.proto

Reply:
I imagine both of them, the common resource names especially, may take more time to implement. So feel free to defer them to follow-up PRs.

Reply:
Thanks! That's absolutely something I did not think about.

Reply:
Filed issues #210 and #209